ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PmtTransactionPool.java

```diff
@@ -0,0 +1,113 @@
+package org.hyperledger.besu.ethereum.privacy;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.chain.BlockAddedEvent;
+import org.hyperledger.besu.ethereum.chain.BlockAddedObserver;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class PmtTransactionPool implements BlockAddedObserver {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Map<Hash, PmtTransactionTracker> pmtPool;
+
+  public PmtTransactionPool() {
+    this.pmtPool = new HashMap<>();
+  }
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent event) {
+    // remove from the map any transactions that went into this block
+    // TODO take into account re-orgs
+    event
+        .getAddedTransactions()
+        .forEach(
+            tx -> {
+              LOG.debug("removing " + pmtPool.containsKey(tx.getHash()));
+              pmtPool.remove(tx.getHash());
+            });
+  }
+
+  public long getMaxMatchingNonce(final String sender, final String privacyGroupId) {
```

Comment:
I think this function would be easier to use when returing an `Optional<Long>`. Right now the cases
- never sent a transaction, hence nonce is 0
- sent a transaction with nonce 0
- sent a transaction with a non-zero nonce but it was confirmed and now we do not know the current nonce

can not be distinguished. Around `nonce==0` this probably leads to wrong behaviour in the caller.

The optional could be used to denote that we do or don't know something about the current nonce.

Reply:
done